pulp_container/tests/functional/api/test_push_content.py

```diff
@@ -4,52 +4,140 @@
+        else:
+            # Ensure logout
+            try:
+                cls.registry.logout(cls.registry_name)
+            except exceptions.CalledProcessError:
+                pass
+
+        cls.registry.pull(image_path)
+
+        if user:
+            cls.registry.logout(cls.registry_name)
+
+    @classmethod
+    def _push(cls, image_path, local_url, user):
+        """
+        Tag and push an image to Pulp registry using specified user.
+        """
         # Tag it to registry under test
-        registry.tag("centos:7", local_url)
-        # Try to push without permission
-        with self.assertRaises(exceptions.CalledProcessError):
-            registry.push(local_url)
+        cls.registry.tag(image_path, local_url)
         # Log in
-        registry.login("-u", "admin", "-p", "password", registry_name)
+        cls.registry.login("-u", user['username'], "-p", user['password'], cls.registry_name)
         # Push successfully
-        registry.push(local_url)
+        cls.registry.push(local_url)
         # Pull while logged in
-        registry.pull(local_url)
-        # Log out
-        registry.logout(registry_name)
+        cls.registry.pull(local_url)
         # Untag local copy
-        registry.rmi(local_url)
-        # Pull while logged out
-        registry.pull(local_url)
+        # TODO: it doesn't seem to work, the tag is still present
+        cls.registry.rmi(local_url)
+
+        cls.registry.logout(cls.registry_name)
+
+    def test_push_using_registry_client_admin(self):
+        """Test push with official registry client and logged in as admin."""
+        local_url = "/".join([self.registry_name, "foo/bar:1.0"])
+        image_path = 'centos:7'
+
+        self._push(image_path, local_url, LOCAL_REGISTRY_ADMIN)
+        self._pull(local_url, LOCAL_REGISTRY_ADMIN)
+
         # cleanup
-        repository = push_repositories_api.list(name="foo/bar").results[0]
-        self.addCleanup(push_repositories_api.delete, repository.pulp_href)
+        distribution = self.distribution_api.list(name="foo/bar").results[0]
+        self.addCleanup(self.distribution_api.delete, distribution.pulp_href)
+
+    def test_push_without_login(self):
+        """Test that one can't push without being logged in."""
+        local_url = "/".join([self.registry_name, "foo/bar:1.0"])
+
+        # Try to push without permission
+        with self.assertRaises(exceptions.CalledProcessError):
+            self.registry.push(local_url)
+
+    def test_push_with_dist_perms(self):
+        """
+        Test that it's enough to have container distribution perms to perform push.
+
+        It also checks read abilities for users with different set of permissions.
+        """
+        repo_name = "test/perms"
+        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
+        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a"
+        self._push(image_path, local_url, self.user_creator)
```

Comment:
this give me an error at the moment
FWIW, I have only podman installed and it's 1.6.4.

```
self = <pulp_container.tests.functional.api.test_push_content.PushRepoTestCase testMethod=test_push_with_dist_perms>

    def test_push_with_dist_perms(self):
        """
        Test that it's enough to have container distribution perms to perform push.
    
        It also checks read abilities for users with different set of permissions.
        """
        repo_name = "test/perms"
        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
        image_path = f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a"
>       self._push(image_path, local_url, self.user_creator)

pulp_container/tests/functional/api/test_push_content.py:115: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pulp_container/tests/functional/api/test_push_content.py:77: in _push
    cls.registry.push(local_url)
../pulp-smash/pulp_smash/cli.py:966: in _dispatch_command
    result = self._client.run(cmd, sudo=True)
../pulp-smash/pulp_smash/cli.py:316: in run
    return self.response_handler(completed_process)
../pulp-smash/pulp_smash/cli.py:54: in code_handler
    completed_proc.check_returncode()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = CompletedProcess(args=('sudo', 'podman', 'push', 'localhost:24817/test/perms:2.0'), returncode=125, stdout='', stderr=...r uploading manifest 2.0 to localhost:24817/test/perms: denied: Access to the requested resource is not authorized.\n')

    def check_returncode(self):
        """Raise an exception if ``returncode`` is non-zero.
    
        Raise :class:`pulp_smash.exceptions.CalledProcessError` if
        ``returncode`` is non-zero.
    
        Why not raise ``subprocess.CalledProcessError``? Because stdout and
        stderr are not included when str() is called on a CalledProcessError
        object. A typical message is::
    
            "Command '('ls', 'foo')' returned non-zero exit status 2"
    
        This information is valuable. One could still make
        ``subprocess.CalledProcessError`` work by overloading ``args``:
    
        >>> if isinstance(args, (str, bytes)):
        ...     custom_args = (args, stdout, stderr)
        ... else:
        ...     custom_args = tuple(args) + (stdout, stderr)
        >>> subprocess.CalledProcessError(args, returncode)
    
        But this seems like a hack.
    
        In addition, it's generally good for an application to raise expected
        exceptions from its own namespace, so as to better abstract away
        dependencies.
        """
        if self.returncode != 0:
            raise exceptions.CalledProcessError(
>               self.args, self.returncode, self.stdout, self.stderr
            )
E           pulp_smash.exceptions.CalledProcessError: Command ('sudo', 'podman', 'push', 'localhost:24817/test/perms:2.0') returned non-zero exit status 125.
E           
E           stdout: 
E           
E           stderr: Getting image source signatures
E           Copying blob sha256:ace0eda3e3be35a979cec764a3321b4c7d0b9e4bb3094d20d3ff6782961a8d54
E           Copying blob sha256:c86ba25590b0e721b9d5a2ca8d2d8e4cd8d694ad2a4235785eb8ac111e95725a
E           Copying config sha256:2ea2d398465fee77f3f80929d6ee391634e548a98eaab5cfb64dba5a25da13b8
E           Writing manifest to image destination
E           Error: Error copying image to the remote destination: Error writing manifest: Error uploading manifest 2.0 to localhost:24817/test/perms: denied: Access to the requested resource is not authorized.

../pulp-smash/pulp_smash/cli.py:129: CalledProcessError

```